src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
+
+				return std::move(params);
+			}();
 
+			std::array<bool, MDI_COMPONENT_COUNT> mdiBytesFilled;
+			std::fill(mdiBytesFilled.data(), mdiBytesFilled.data() + MDI_COMPONENT_COUNT, false);
+			std::array<MDI_SIZE_TYPE, MDI_COMPONENT_COUNT> mdiOffsets;
+			std::fill(mdiOffsets.data(), mdiOffsets.data() + MDI_COMPONENT_COUNT, INVALID_ADDRESS);
+
+			auto streamingBuffer = m_mdi.buffer;
+			{
+				auto binding = streamingBuffer->getBuffer()->getBoundMemory();
+				assert(binding.memory->isCurrentlyMapped());
+
+				auto* const mdiData = reinterpret_cast<uint8_t*>(streamingBuffer->getBufferPointer());
+
+				struct
 				{
-					if (!iBinding.memory->map({ 0ull, iBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for index buffer data!", system::ILogger::ELL_WARNING);
-				
-					assert(iBinding.memory->isCurrentlyMapped());
-				}
+					MDI_SIZE_TYPE offset = INVALID_ADDRESS,
+					multiAllocationSize = {};
+				} requestState;
 
-				auto* vertex_ptr = static_cast<ImDrawVert*>(vBinding.memory->getMappedPointer());
-				auto* index_ptr = static_cast<ImDrawIdx*>(iBinding.memory->getMappedPointer());
+				const auto start = std::chrono::steady_clock::now();
 
-				for (int n = 0; n < drawData->CmdListsCount; n++)
+				//! we must upload entire MDI buffer data to our streaming buffer, but we cannot guarantee allocation can be done in single request
+				for (MDI_SIZE_TYPE uploadedSize = 0ull; uploadedSize < mdiParams.totalByteSizeRequest;)
 				{
-					const ImDrawList* cmd = drawData->CmdLists[n];
-					::memcpy(vertex_ptr, cmd->VtxBuffer.Data, cmd->VtxBuffer.Size * sizeof(ImDrawVert));
-					::memcpy(index_ptr, cmd->IdxBuffer.Data, cmd->IdxBuffer.Size * sizeof(ImDrawIdx));
-					vertex_ptr += cmd->VtxBuffer.Size;
-					index_ptr += cmd->IdxBuffer.Size;
-				}
+					auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start);
+
+					if (elapsed.count() >= 1u)
+					{
+						streamingBuffer->cull_frees();
+						return false;
+					}
+
+					requestState.offset = INVALID_ADDRESS;
+					requestState.multiAllocationSize = streamingBuffer->max_size(); // request available block memory size which we can try to allocate
+
+					static constexpr auto STREAMING_ALLOCATION_COUNT = 1u;
+					const size_t unallocatedSize = m_mdi.buffer->multi_allocate(std::chrono::steady_clock::now() + std::chrono::microseconds(100u), STREAMING_ALLOCATION_COUNT, &requestState.offset, &requestState.multiAllocationSize, &MDI_MAX_ALIGNMENT); //! (*) note we request single tight chunk of memory with max alignment instead of MDI::E_BUFFER_CONTENT separate chunks 
 
-				vBinding.memory->unmap();
-				iBinding.memory->unmap();
+					if (requestState.offset == INVALID_ADDRESS)
+						continue; // failed? lets try again, TODO: should I here have my "blockMemoryFactor =* 0.5" and apply to requestState.multiAllocationSize?
```

Comment:
yes, but it should be a minimum of previous `max_size` and 0.5 of the last size you tried

(then obvs initialize `multuAllocationSize` to `mdiParams.totalByteSizeRequest` before the loop) 
Also in the success case, you need to re-initialize to `multiAllocationSize` to the bytes left to upload